lib/Dialect/FIRRTL/Transforms/LegacyWiring.cpp

```diff
@@ -0,0 +1,120 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the legacy Wiring annotation resolver.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Support/LogicalResult.h"
+
+using namespace circt;
+using namespace firrtl;
+
+/// Consume SourceAnnotation and SinkAnnotation, storing into state
+LogicalResult circt::firrtl::applyWiring(const AnnoPathValue &target,
+                                         DictionaryAttr anno,
+                                         ApplyState &state) {
+  auto *context = anno.getContext();
+  ImplicitLocOpBuilder builder(target.ref.getOp()->getLoc(), context);
+
+  // Convert target to Value
+  SmallVector<Value> targetsValues;
+  if (auto portTarget = target.ref.dyn_cast<PortAnnoTarget>()) {
+    auto portNum = portTarget.getImpl().getPortNo();
+    if (auto module = dyn_cast<FModuleOp>(portTarget.getOp())) {
+      builder.setInsertionPointToEnd(module.getBodyBlock());
+      targetsValues.push_back(getValueByFieldID(
+          builder, module.getArgument(portNum), target.fieldIdx));
+    } else if (auto ext = dyn_cast<FExtModuleOp>(portTarget.getOp())) {
+      if (target.instances.empty()) {
+        auto paths = state.instancePathCache.getAbsolutePaths(ext);
+        if (paths.size() > 1) {
+          mlir::emitError(state.circuit.getLoc())
+              << "cannot resolve a unique instance path from the "
+                 "external module target"
+              << target.ref;
+          return failure();
+        }
+        auto inst = cast<InstanceOp>(paths[0].back());
+        builder.setInsertionPointAfter(inst);
+        targetsValues.push_back(getValueByFieldID(
+            builder, inst->getResult(portNum), target.fieldIdx));
+      } else {
+        auto inst = cast<InstanceOp>(target.instances.back());
+        builder.setInsertionPointAfter(inst);
+        targetsValues.push_back(getValueByFieldID(
+            builder, inst->getResult(portNum), target.fieldIdx));
+      }
+    } else {
+      return mlir::emitError(state.circuit.getLoc())
+             << "Annotation has invalid target: " << anno;
+    }
+  } else if (auto opResult = target.ref.dyn_cast<OpAnnoTarget>()) {
+    if (target.isOpOfType<WireOp, RegOp, RegResetOp>()) {
+      auto module = cast<FModuleOp>(opResult.getModule());
+      builder.setInsertionPointToEnd(module.getBodyBlock());
+      if (opResult.getOp()->getNumResults() != 1) {
+        // Unsure about what ops emit more/less than 1 result, perhaps the type
```

Comment:
Yep, all of these operation types only have a single result and I don't think that'll change.